merkletree/proof.go

```diff
@@ -16,6 +23,75 @@ type ProofNode interface {
 	Commitment() []byte
 }
 
+func computeLeafHash(ap *AuthenticationPath,
+	leafIndex, leafCommitment []byte, leafLevel int, isLeafEmpty bool) (leafHash []byte) {
+	if isLeafEmpty {
```

Comment:
> Or is your point that the client doesn't have to re-serialize the AP (like for a signature in the case of the STR)?

Yes, it is what I mean. I think what we're doing here is just packing inputs into a struct and pass it into the verification function, instead of passing a lot of parameters. Maybe it also has some advantages in performance (e.g., smaller stack frame ...) :D .